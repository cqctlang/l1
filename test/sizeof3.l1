define listref(l, i){
	if(i == 0)
		return car(l);
	else
		return listref(cdr(l), i-1);
}

define mkzas(len){
	@local dispatch, str;
	str = mkstr(len);
	define dispatch args {
		@local arg, range, s, new, beg, end;
		arg = listref(args, 0);

		if(arg == "get"){
			range = listref(args, 1);
			beg = rangebeg(range);
			end = beg+rangelen(range);
			printf("get %x %x\n", beg, end);
			return substr(str, beg, end);
		}

		if(arg == "put"){
			range = listref(args, 1);
			beg = rangebeg(range);
			end = beg+rangelen(range);
			new = listref(args, 2);
			strput(str, beg, new);
			printf("put %x %x\n", beg, end);
			return;
		}

		printf("bad argument to dispatch!\n");
	}
	return mkas(dispatch);
}

define main(){
	@local dom, ns, as, x, y;

	as = mkzas(1024);
	ns = @names c32le {
		struct foo {
			@0x0 int i;
			@0x4 int *p;
			@0x8;
		};
		@0x0	int x;
		@0x4	int y;
		@0x8	struct foo foo;
		@0x10	struct foo a[3];
	};
	dom = mkdom(ns, as);

	printf("sizeof(dom`int) == %d\n", sizeof(dom`int));
	printf("sizeof(struct dom`foo) == %d\n", sizeof(struct dom`foo));
	printf("sizeof(struct dom`foo*) == %d\n", sizeof(struct dom`foo*));

	printf("sizeof(3) == %d\n", sizeof(3));
	printf("sizeof(dom`a) == %d\n", sizeof(dom`a));	
	printf("sizeof(dom`a[2]) == %d\n", sizeof(dom`a[2]));
	printf("sizeof(dom`a[2].i) == %d\n", sizeof(dom`a[2].i));
}

main();
