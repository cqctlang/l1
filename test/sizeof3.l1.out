entry: var params, 1 locals, 0 temps
listref: 2 params, 0 locals, 0 temps
mkzas: 1 params, 3 locals, 0 temps
dispatch: var params, 7 locals, 0 temps
main: 0 params, 17 locals, 0 temps
sizeof(dom`int) == 4
sizeof(struct dom`foo) == 8
sizeof(struct dom`foo*) == 4
sizeof(3) == 4
sizeof(dom`a) == 24
sizeof(dom`a[2]) == 8
sizeof(dom`a[2].i) == 4
